crates/polars-ops/src/frame/unpivot.rs

```diff
@@ -66,73 +67,55 @@ pub trait UnpivotDF: IntoDf {
     ///  +-----+-----+----------+-------+
     ///  | "a" | 5   | "D"      | 6     |
     ///  +-----+-----+----------+-------+
+    ///
+    /// The resulting row order is unspecified.
     /// ```
-    fn unpivot<I, J>(&self, on: I, index: J) -> PolarsResult<DataFrame>
+    fn unpivot<I, J>(&self, on: Option<I>, index: J) -> PolarsResult<DataFrame>
     where
         I: IntoVec<PlSmallStr>,
         J: IntoVec<PlSmallStr>,
     {
-        let index = index.into_vec();
-        let on = on.into_vec();
-        self.unpivot2(UnpivotArgsIR {
-            on,
-            index,
-            ..Default::default()
-        })
+        self.unpivot2(UnpivotArgsIR::new(
+            self.to_df().get_column_names_owned(),
+            on.map(|on| on.into_vec()),
+            index.into_vec(),
+            None,
+            None,
+        ))
     }
 
     /// Similar to unpivot, but without generics. This may be easier if you want to pass
     /// an empty `index` or empty `on`.
     fn unpivot2(&self, args: UnpivotArgsIR) -> PolarsResult<DataFrame> {
+        let UnpivotArgsIR {
+            on,
+            index,
+            variable_name,
+            value_name,
+        } = args;
+
         let self_ = self.to_df();
-        let index = args.index;
-        let mut on = args.on;
 
-        let variable_name = args
-            .variable_name
-            .unwrap_or_else(|| PlSmallStr::from_static("variable"));
-        let value_name = args
-            .value_name
-            .unwrap_or_else(|| PlSmallStr::from_static("value"));
+        let variable_col_empty = Column::new_empty(variable_name.clone(), &DataType::String);
+        let value_col_empty = Column::new_empty(value_name.clone(), &DataType::Null);
 
         if self_.get_columns().is_empty() {
-            return DataFrame::new(vec![
-                Column::new_empty(variable_name, &DataType::String),
-                Column::new_empty(value_name, &DataType::Null),
-            ]);
+            return DataFrame::new(vec![variable_col_empty, value_col_empty]);
         }
 
-        let len = self_.height();
+        // If the parameter `on` is empty or there are no columns available to use as value vars. we
+        // want to produce an empty DataFrame but with the standard unpivot schema.
+        if on.is_empty() || index.len() == self_.width() {
```

Comment:
I don't think we should hit this branch for `index.len() == self_.width()`. Columns can be part of both `on` and `index`:

```python
>>> data = {
...     "a": ["x", "y", "z"],
...     "b": [1, 3, 5],
...     "c": [2, 4, 6],
...     "d": [77, 88, 99],
... }
...
... print(pl.DataFrame(data).unpivot(["a"], index=["a", "b", "c", "d"]))
shape: (3, 6)
┌─────┬─────┬─────┬─────┬──────────┬───────┐
│ a   ┆ b   ┆ c   ┆ d   ┆ variable ┆ value │
│ --- ┆ --- ┆ --- ┆ --- ┆ ---      ┆ ---   │
│ str ┆ i64 ┆ i64 ┆ i64 ┆ str      ┆ str   │
╞═════╪═════╪═════╪═════╪══════════╪═══════╡
│ x   ┆ 1   ┆ 2   ┆ 77  ┆ a        ┆ x     │
│ y   ┆ 3   ┆ 4   ┆ 88  ┆ a        ┆ y     │
│ z   ┆ 5   ┆ 6   ┆ 99  ┆ a        ┆ z     │
└─────┴─────┴─────┴─────┴──────────┴───────┘
```